polkadot/runtime/parachains/src/paras_inherent/mod.rs

```diff
@@ -1166,5 +1211,103 @@ fn filter_backed_statements_from_disabled_validators<T: shared::Config + schedul
 	});
 
 	// Also return `true` if a whole candidate was dropped from the set
-	filtered || backed_len_before != backed_candidates.len()
+	filtered || backed_len_before != backed_candidates_with_core.len()
+}
+
+/// Map candidates to scheduled cores.
+/// If the para only has one scheduled core and no `CoreIndex` is injected, map the candidate to the
+/// single core. If the para has multiple cores scheduled, only map the candidates which have a
+/// proper core injected. Filter out the rest.
+fn map_candidates_to_cores<T: configuration::Config + scheduler::Config + inclusion::Config>(
+	allowed_relay_parents: &AllowedRelayParentsTracker<T::Hash, BlockNumberFor<T>>,
+	mut scheduled: BTreeMap<ParaId, BTreeSet<CoreIndex>>,
+	core_index_enabled: bool,
+	candidates: Vec<BackedCandidate<T::Hash>>,
+) -> (Vec<(BackedCandidate<T::Hash>, CoreIndex)>, bool) {
+	let mut dropped_elastic_scaling_candidates = false;
+	let mut backed_candidates_with_core = Vec::with_capacity(candidates.len());
+
+	// We keep a candidate if the parachain has only one core assigned or if
+	// a core index is provided by block author and it's indeed scheduled.
+	for backed_candidate in candidates {
+		let maybe_injected_core_index = get_injected_core_index::<T>(
+			allowed_relay_parents,
+			&backed_candidate,
+			core_index_enabled,
+		);
+
+		let scheduled_cores = scheduled.get_mut(&backed_candidate.descriptor().para_id);
+		if let Some(scheduled_cores) = scheduled_cores {
```

Comment:
```suggestion
        // Candidates without scheduled cores are silently filtered out.
		if let Some(scheduled_cores) = scheduled_cores {
```

I'd make this filtering more explicit but this is a very personal preference.

But let's have a comment pointing out why there is no `else` here.